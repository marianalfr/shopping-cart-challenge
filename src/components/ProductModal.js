import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { Button, AddRemoveButton, Quantity } from './Elements';

// Styled Components -->

const ModalWrapper = styled.div`
  position: fixed;
  top: 50%;
  left: 50%;
  display: flex;
  overflow-x: hidden;
  overflow-y: auto;
  max-width: 1088px;
  max-height: 648px;
  width: 100%;
  height: 100%;
  border-radius: 4px;
  background-color: ${props => props.theme.color.lightest};
  transform: translate(-50%, -50%);
  z-index: 2;
`;

const ImageWrapper = styled.div`
    height: 100%;
    flex: 1;
    background-image: ${props => `url(${props.product.images.large})`};
    background-size: cover;
    background-position: center;
`;

const InfoWrapper = styled.aside`
  display: flex;
  flex-flow: column wrap;
  padding: 32px;
  width: 312px;
  color: ${props => props.theme.color.mainDark};
`;

const Close = styled.div`
    width: 100%;
    text-align: right;
    align-self: flex-end;
    cursor: pointer;
    font-size: 25px;
    color: ${props => props.theme.color.light};
`;

const Info = styled.div`
    width: 100%;
    flex: 2;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    align-items: center;
`;

const Title = styled.h1`
    width: 100%;
    padding-bottom: 16px;
    border-bottom: ${props => props.theme.border.bottom};
    display: flex;
    justify-content: space-between;
    font-size: 18px;
    font-weight: bold;
    line-height: 17px;
    text-transform: capitalize;
`;

const Description = styled.p`
    padding: 16px 0 32px 0;
    border-bottom: ${props => props.theme.border.bottom};
    font-size: 14px;
`;

const Offer = styled.p`
    width: 100%;
    text-align: left;
    color: ${props => props.theme.color.main};
    padding: 16px 0;
    border-bottom: ${props => props.theme.border.bottom};
    font-weight: 400;
    font-size: 14px;
    line-height: 16px;
`;

const Code = styled.p`
    padding: 12px 0;
    color: ${props => props.theme.color.light};
    letter-spacing: 0.13px;
    font-weight: 400;
    font-size: 12px;
    line-height: 13px;
    align-self: flex-start;
`;

const Cart = styled.div`
    flex: 1;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
`;

const Bold = styled.p`
    padding: 16px 0;
    font-size: 14px;
    font-weight: bold;
    align-self: flex-start;
    
`;

const AddRemove = styled.div`
    display: flex;
`;

// ProductModal Component ------------------------------------------------------------->

const ProductModal = props => {
    return(
        <ModalWrapper>
            <ImageWrapper product = {props.product}></ImageWrapper>
            <InfoWrapper>
                <Close onClick={props.toggleModal}>✕</Close>
                <Info>
                    <Title>
                        <span>{props.product.product}</span>
                        <span>{props.product.price}€</span>
                    </Title>
                    {props.product.offer.type !== null ? (
                    <Offer>Offer: {props.product.offer.type} when you buy {props.product.offer.minQty} or more.</Offer>
                    ) : ''}
                    <Description>{props.product.description}</Description>
                    <Code>Product code: {props.product.code}</Code>
                </Info>
                <Cart>
                    {props.product.quantity !== 0 ? (
                        <React.Fragment>
                            <Bold>Currently in your shopping cart:</Bold>
                            <AddRemove>
                                <AddRemoveButton onClick={props.removeItem}>-</AddRemoveButton>
                                <Quantity type="text" value={props.product.quantity} readOnly></Quantity>
                                <AddRemoveButton onClick={props.addItem}>+</AddRemoveButton>
                            </AddRemove>
                        </React.Fragment>
                    ) : (
                        <Button type="button" onClick={props.addItem}>Add to cart</Button>
                    )}
                </Cart>
            </InfoWrapper>
        </ModalWrapper>
    );
};

ProductModal.propTypes = {
    product:PropTypes.shape({
        product: PropTypes.string,
        price: PropTypes.number,
        quantity: PropTypes.number,
        code: PropTypes.string,
        description: PropTypes.string,
        images: PropTypes.shape({
            thumb: PropTypes.string,
            large: PropTypes.string
        }),
        offer: PropTypes.shape({
            type: PropTypes.string,
            minQty: PropTypes.number 
        })
    }),
    toggleModal: PropTypes.func,
    addItem: PropTypes.func,
    removeItem: PropTypes.func
};

export default ProductModal;