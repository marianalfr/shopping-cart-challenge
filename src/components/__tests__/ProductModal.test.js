import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import 'jest-styled-components';
import Adapter from 'enzyme-adapter-react-16';
import ProductModal from '../ProductModal';

Enzyme.configure({ adapter: new Adapter() });

const item = {
    product: {
      name: 'shirt',
      price: 20,
      code: 'X7R2OPX',
      description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
      images: {
        thumb: require('../../images/shirt.png'),
        large: require('../../images/shirt-large.jpg')
      },
      offer: {
        type: {
          category: 'percentage',
          name: '-5%',
          numbers: {
            percentage: 5
          }
        },
        minQty: 3 
      }
    },
    quantity: 3
  };

describe('ProductModal Component', () => {

    it('renders correctly', () => {
        const wrapper = shallow(
            <ProductModal 
                item = { item }
                toggleModal = { () => {} }
                addItem = { () => {} }
                removeItem = { () => {} }
            />
        );
        expect(wrapper.debug()).toMatchSnapshot(); 
    });
});