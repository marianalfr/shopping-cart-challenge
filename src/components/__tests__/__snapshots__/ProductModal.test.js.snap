// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductModal Component renders correctly 1`] = `
"<styled.div>
  <styled.div product={{...}} />
  <styled.aside>
    <styled.div onClick={[Function: toggleModal]}>
      ✕
    </styled.div>
    <styled.div>
      <styled.h1>
        <span>
          shirt
        </span>
        <span>
          20
          €
        </span>
      </styled.h1>
      <styled.p>
        Offer: 
        -5%
         when you buy 
        3
         or more.
      </styled.p>
      <styled.p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
      </styled.p>
      <styled.p>
        Product code: 
        X7R2OPX
      </styled.p>
    </styled.div>
    <styled.div>
      <styled.p>
        Currently in your shopping cart:
      </styled.p>
      <styled.div>
        <styled.button onClick={[Function: removeItem]}>
          -
        </styled.button>
        <styled.input type=\\"text\\" value={3} readOnly={true} />
        <styled.button onClick={[Function: addItem]}>
          +
        </styled.button>
      </styled.div>
    </styled.div>
  </styled.aside>
</styled.div>"
`;
